src/daft-csv/src/read.rs

```diff
@@ -0,0 +1,290 @@
+                File::open(file.path).await?.compat(),
+                column_names,
+                include_columns,
+                num_rows,
+                has_header,
+                delimiter,
+            )
+            .await
+        }
+        result @ GetResult::Stream(..) => {
+            // TODO(Clark): Enable streaming remote reads by wrapping the BoxStream in a buffered stream that's
+            // (1) sync and (2) seekable.
+            read_csv_single_from_reader(
+                Cursor::new(result.bytes().await?),
+                column_names,
+                include_columns,
+                num_rows,
+                has_header,
+                delimiter,
+            )
+            .await
+        }
+    }
+}
+
+async fn read_csv_single_from_reader<R>(
+    reader: R,
+    column_names: Option<Vec<&str>>,
+    include_columns: Option<Vec<&str>>,
+    num_rows: Option<usize>,
+    has_header: bool,
+    delimiter: Option<u8>,
+) -> DaftResult<Table>
+where
+    R: AsyncRead + AsyncSeek + Unpin + Sync + Send,
+{
+    let mut reader = AsyncReaderBuilder::new()
+        .has_headers(has_header)
+        .delimiter(delimiter.unwrap_or(b','))
+        .create_reader(reader);
+    let (mut fields, _) = infer_schema(&mut reader, None, has_header, &infer).await?;
+    if let Some(column_names) = column_names {
+        fields = fields
+            .into_iter()
+            .zip(column_names.iter())
+            .map(|(field, name)| {
+                Field::new(*name, field.data_type, field.is_nullable).with_metadata(field.metadata)
+            })
+            .collect();
+    }
+    let field_name_to_idx = fields
+        .iter()
+        .enumerate()
+        .map(|(idx, f)| (f.name.as_ref(), idx))
+        .collect::<HashMap<&str, usize>>();
+    let projection_indices = include_columns.as_ref().map(|cols| {
+        cols.iter()
+            .map(|c| field_name_to_idx[c])
+            .collect::<Vec<_>>()
+    });
+    let num_rows = num_rows.unwrap_or(usize::MAX);
+    // TODO(Clark): Make batch size configurable.
+    // TODO(Clark): Get estimated average row size in bytes during schema inference and use it to:
+    //   1. Set a reasonable batch size.
+    //   2. Preallocate per-column batch vecs.
+    //   3. Preallocate column Arrow array buffers.
+    let batch_size = 1024.min(num_rows);
+    // TODO(Clark): Instead of allocating an array-per-column-batch and concatenating at the end,
+    // progressively grow a single array per column (with the above preallocation based on estimated
+    // number of rows).
+    let mut column_arrays = vec![
+        vec![];
+        projection_indices
+            .as_ref()
+            .map(|p| p.len())
+            .unwrap_or(fields.len())
+    ];
+    let mut buffer = vec![ByteRecord::with_capacity(0, fields.len()); batch_size];
+    let mut rows = buffer.as_mut_slice();
+    // Number of rows read in last read.
+    let mut rows_read = 1;
```

Comment:
is this always true or only when theres a header?

Reply:
This is actually orthogonal to the presence of a header; we just need to set this to some positive number to ensure that we can enter the while loop (i.e. the condition `rows_read > 0` passes). `rows_read` will get overwritten by the return value of `read_rows()` in the while loop body, before `rows_read` is used for anything other than the while loop condition.

If this is confusing, we could set `rows_read = -1` and change the while loop condition to `rows_read != 0`, if that makes the semantics of the initial `rows_read` value more clear?